Feature: Handle "optional" imports with try/catch

Node.js libraries sometimes use a pattern of "conditionally requiring" code like:

```js
try {
  const isValidUTF8 = require('utf-8-validate');
  exports.isValidUTF8 =
    typeof isValidUTF8 === 'object'
      ? isValidUTF8.Validation.isValidUTF8 // utf-8-validate@<3.0.0
      : isValidUTF8;
} catch (e) /* istanbul ignore next */ {
  exports.isValidUTF8 = () => true;
}
```

See, e.g.:

- https://github.com/websockets/ws/blob/b6430fea423d88926847a47d4ecfc36e52dc1164/lib/validation.js#L3-L10
- https://github.com/websockets/ws/blob/08c6c8ba70404818f7f4bc23eb5fd0bf9c94c039/lib/buffer-util.js#L121-L122

... then a project could legitimately not have `utf-8-validate` or `bufferutil` installed. But, `trace-deps` still errors with:

```
Error: Encountered resolution error in /PATH/TO/PROJ/node_modules/ws/lib/buffer-util.js for bufferutil: Error: Cannot find module 'bufferutil' from '/PATH/TO/PROJ//node_modules/ws/lib'
      at resolve.then.catch (/PATH/TO/PROJ/node_modules/trace-deps/lib/trace.js:105:17)
```

We don't want to just do `ignores` because _other_ packages might have one of those packages locally discoverable.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,8 @@ Changes
 
 ## 0.2.2
 
+* Bug/Feature: Allow permissive handling of try/catch `require`s.
+  [#19](https://github.com/FormidableLabs/trace-deps/issues/19)
 * Upgrade `node-acorn` to `^2.0.0`.
 
 ## 0.2.1

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,7 +3,7 @@ Changes
 
 ## 0.2.2
 
-* Bug/Feature: Allow permissive handling of try/catch `require`s.
+* Bug/Feature: Allow permissive handling of try/catch `require`s with `allowMissing` parameter.
   [#19](https://github.com/FormidableLabs/trace-deps/issues/19)
 * Upgrade `node-acorn` to `^2.0.0`.
 

--- a/README.md
+++ b/README.md
@@ -28,7 +28,7 @@ _Parameters_:
 
 * `srcPath` (`string`): source file path to trace
 * `ignores` (`Array<string>`): list of package prefixes to ignore tracing entirely
-* `allowMissing` (`Object<string, Array<string>`): Mapping of package prefixes to permitted
+* `allowMissing` (`Object.<string, Array<string>`): Mapping of package prefixes to permitted
   missing module prefixes.
 
 _Returns_:
@@ -43,7 +43,7 @@ _Parameters_:
 
 * `srcPaths` (`Array<string>`): source file paths to trace
 * `ignores` (`Array<string>`): list of package prefixes to ignore
-* `allowMissing` (`Object<string, Array<string>`): Mapping of package prefixes to permitted
+* `allowMissing` (`Object.<string, Array<string>`): Mapping of package prefixes to permitted
   missing module prefixes.
 
 _Returns_:
@@ -57,6 +57,24 @@ _Returns_:
 * **Only handles single string dependencies**: `require`, `require.resolve`, and dynamic `import()` support calls with variables or other expressions like `require(aVar)`, `import(process.env.VAL + "more-stuff")`. This library presently only supports calls with a **single string** and nothing else. We have a [tracking ticket](https://github.com/FormidableLabs/trace-deps/issues/2) to consider expanding support for things like partial evaluation.
 
 * **Includes `package.json` files used in resolution**: As this is a Node.js-focused library, to follow the Node.js [module resolution algorithm](https://nodejs.org/api/modules.html#modules_all_together) which notably uses intermediate encountered `package.json` files to determine how to resolve modules. This means that we include a lot of `package.json` files that seemingly aren't directly imported (such as a `const pkg = require("mod/package.json")`) because they are needed for the list of all traced files to together resolve correctly if all on disk together.
+
+* **Using the `allowMissing` option**: The `allowMissing` function field helps in situations where you want to allow certain dependencies to have known missing sub-dependencies, often seen in patterns like: `try { require("optional-dep"); } catch (e) {}`. If the sub-dependency is found, then it will be returned just like any normal one. If not, the module not found error is just swallowed and normal processing resumes.
+
+    To configure the parameter, create an object of key `package-prefix` with a value of an array of other package prefixes to skip over not found errors:
+
+    ```js
+    traceFile({
+      srcPath,
+      allowMissing: {
+        "ws": [
+          // See, e.g.: https://github.com/websockets/ws/blob/08c6c8ba70404818f7f4bc23eb5fd0bf9c94c039/lib/buffer-util.js#L121-L122
+          "bufferutil",
+          // See, e.g.: https://github.com/websockets/ws/blob/b6430fea423d88926847a47d4ecfc36e52dc1164/lib/validation.js#L3-L10
+          "utf-8-validate"
+        ]
+      }
+    })
+    ```
 
 [npm_img]: https://badge.fury.io/js/trace-deps.svg
 [npm_site]: http://badge.fury.io/js/trace-deps

--- a/README.md
+++ b/README.md
@@ -27,7 +27,9 @@ Trace and return on-disk locations of all file dependencies from a source file.
 _Parameters_:
 
 * `srcPath` (`string`): source file path to trace
-* `ignores` (`Array<string>`): list of package prefixes to ignore
+* `ignores` (`Array<string>`): list of package prefixes to ignore tracing entirely
+* `allowMissing` (`Object<string, Array<string>`): Mapping of package prefixes to permitted
+  missing module prefixes.
 
 _Returns_:
 
@@ -41,6 +43,8 @@ _Parameters_:
 
 * `srcPaths` (`Array<string>`): source file paths to trace
 * `ignores` (`Array<string>`): list of package prefixes to ignore
+* `allowMissing` (`Object<string, Array<string>`): Mapping of package prefixes to permitted
+  missing module prefixes.
 
 _Returns_:
 

--- a/README.md
+++ b/README.md
@@ -76,6 +76,8 @@ _Returns_:
     })
     ```
 
+* **`ignores` vs. `allowMissing`**: The `ignores` option completely skips a dependency from being further traversed irrespective of whether or not a matching dependency exists on disk. The `allowMissing` option will include and further traverse dependencies that are present on disk if found, and suppress any errors upon hitting a matching missing dependency.
+
 [npm_img]: https://badge.fury.io/js/trace-deps.svg
 [npm_site]: http://badge.fury.io/js/trace-deps
 [trav_img]: https://api.travis-ci.com/FormidableLabs/trace-deps.svg

--- a/README.md
+++ b/README.md
@@ -76,7 +76,7 @@ _Returns_:
     })
     ```
 
-* **`ignores` vs. `allowMissing`**: The `ignores` option completely skips a dependency from being further traversed irrespective of whether or not a matching dependency exists on disk. The `allowMissing` option will include and further traverse dependencies that are present on disk if found, and suppress any errors upon hitting a matching missing dependency.
+* **`ignores` vs. `allowMissing`**: The `ignores` option completely skips a dependency from being further traversed irrespective of whether or not a matching dependency exists on disk. The `allowMissing` option will include and further traverse dependencies that are present on disk if found and suppress any errors for matches that are missing.
 
 [npm_img]: https://badge.fury.io/js/trace-deps.svg
 [npm_site]: http://badge.fury.io/js/trace-deps

--- a/lib/package.js
+++ b/lib/package.js
@@ -7,10 +7,10 @@ const path = require("path");
 
 // Return in-order list of all `node_modules` packages from a file path.
 const getPackages = (filePath) => {
-  if (!filePath || !filePath.length) { return []; }
+  if (!filePath) { return []; }
 
-  // Iterate all normalized parts of the file path.
-  const pkgs = path.normalize(filePath).split(path.sep)
+  // Iterate all normalized parts of the file path and extract packages.
+  return path.normalize(filePath).split(path.sep)
     .map((part, i, parts) => {
       if (parts[i - 1] === "node_modules") {
         if (part[0] !== "@") {
@@ -28,8 +28,6 @@ const getPackages = (filePath) => {
       return null;
     })
     .filter(Boolean);
-
-  return pkgs;
 };
 
 // Return last (deepest) package in a file path.

--- a/lib/package.js
+++ b/lib/package.js
@@ -32,6 +32,10 @@ const getPackages = (filePath) => {
   return pkgs;
 };
 
+// Return last (deepest) package in a file path.
+const getLastPackage = (filePath) => getPackages(filePath).pop() || null;
+
 module.exports = {
-  getPackages
+  getPackages,
+  getLastPackage
 };

--- a/lib/package.js
+++ b/lib/package.js
@@ -1,0 +1,37 @@
+"use strict";
+
+/**
+ * Package utilities.
+ */
+const path = require("path");
+
+// Return in-order list of all `node_modules` packages from a file path.
+const getPackages = (filePath) => {
+  if (!filePath || !filePath.length) { return []; }
+
+  // Iterate all normalized parts of the file path.
+  const pkgs = path.normalize(filePath).split(path.sep)
+    .map((part, i, parts) => {
+      if (parts[i - 1] === "node_modules") {
+        if (part[0] !== "@") {
+          // Unscoped: Add name.
+          return part;
+        }
+
+        const nextPart = parts[i + 1];
+        if (nextPart) {
+          // Scoped: Add scope + name.
+          return `${part}/${nextPart}`;
+        }
+      }
+
+      return null;
+    })
+    .filter(Boolean);
+
+  return pkgs;
+};
+
+module.exports = {
+  getPackages
+};

--- a/lib/trace.js
+++ b/lib/trace.js
@@ -101,7 +101,12 @@ const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } =
         // Add in package files if any.
         .then((depPath) => ({ depPath, pkgPaths }))
         .catch((err) => {
-        // Convert to more useful error.
+          // Check if allowed to be missing.
+          if (err.code === "MODULE_NOT_FOUND") {
+            console.log("TODO HERE MISSING", { srcPath, depName });
+          }
+
+          // Convert to more useful error.
           throw new Error(`Encountered resolution error in ${srcPath} for ${depName}: ${err}`);
         });
     }))

--- a/lib/trace.js
+++ b/lib/trace.js
@@ -52,11 +52,12 @@ const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, tracedDepPaths }
  * point file you should manually create and pass a shared `tracedDepPaths` set
  * to avoid unnecessary additional tracing.
  *
- * @param {*}                         opts                options object
- * @param {string}                    opts.srcPath        source file path to trace
- * @param {Array<string>}             opts.ignores        list of package prefixes to ignore
- * @param {Set}                       opts.tracedDepPaths tracked dependencies
- * @returns {Promise<Array<string>>}  list of absolute paths to on-disk dependencies
+ * @param {*}             opts                options object
+ * @param {string}        opts.srcPath        source file path to trace
+ * @param {Array<string>} opts.ignores        list of package prefixes to ignore
+ * @param {Object}        opts.allowMissing   map packages to list of allowed missing package
+ * @param {Set}           opts.tracedDepPaths tracked dependencies
+ * @returns {Promise<Array<string>>}          list of absolute paths to on-disk dependencies
  */
 const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } = {}) => {
   if (!srcPath) {
@@ -133,11 +134,12 @@ const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } =
 /**
  * Trace and return on-disk locations of all file dependencies from source files.
  *
- * @param {*}                         opts                options object
- * @param {Array<string>}             opts.srcPaths       source file paths to trace
- * @param {Array<string>}             opts.ignores        list of package prefixes to ignore
- * @param {Set}                       opts.tracedDepPaths tracked dependencies
- * @returns {Promise<Array<string>>}  list of absolute paths to on-disk dependencies
+ * @param {*}             opts                options object
+ * @param {Array<string>} opts.srcPaths       source file paths to trace
+ * @param {Array<string>} opts.ignores        list of package prefixes to ignore
+ * @param {Object}        opts.allowMissing   map packages to list of allowed missing package
+ * @param {Set}           opts.tracedDepPaths tracked dependencies
+ * @returns {Promise<Array<string>>}          list of absolute paths to on-disk dependencies
  */
 const traceFiles = async ({ srcPaths, ignores = [], tracedDepPaths = new Set() } = {}) => {
   // Recurse all source files.

--- a/lib/trace.js
+++ b/lib/trace.js
@@ -6,6 +6,7 @@ const { promisify } = require("util");
 
 const { parse } = require("acorn-node");
 const { getDeps } = require("./extract");
+const { getLastPackage } = require("./package");
 
 // Helpers
 const resolve = promisify(require("resolve"));
@@ -15,7 +16,7 @@ const uniq = (items) => Array.from(new Set(items)).sort();
 // HELPER: Recursively track and trace dependency paths.
 // - `srcPaths`: If provided, **don't** add to resolvedDepPaths
 // - `depPaths`: If provided, **do** add to resolvedDepPaths
-const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, tracedDepPaths }) => {
+const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, allowMissing, tracedDepPaths }) => {
   // Start **only** with depPaths.
   let resolvedDepPaths = Array.from(depPaths);
 
@@ -32,7 +33,7 @@ const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, tracedDepPaths }
 
       // Recurse.
       // eslint-disable-next-line no-use-before-define
-      const recursedDepPaths = await traceFile({ srcPath: depPath, ignores, tracedDepPaths });
+      const recursedDepPaths = await traceFile({ srcPath: depPath, ignores, allowMissing, tracedDepPaths });
 
       // Aggregate.
       resolvedDepPaths = resolvedDepPaths.concat(recursedDepPaths);
@@ -59,7 +60,7 @@ const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, tracedDepPaths }
  * @param {Set}           opts.tracedDepPaths tracked dependencies
  * @returns {Promise<Array<string>>}          list of absolute paths to on-disk dependencies
  */
-const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } = {}) => {
+const traceFile = async ({ srcPath, ignores = [], allowMissing = {}, tracedDepPaths = new Set() } = {}) => {
   if (!srcPath) {
     throw new Error("Empty source file path");
   }
@@ -104,15 +105,26 @@ const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } =
         .catch((err) => {
           // Check if allowed to be missing.
           if (err.code === "MODULE_NOT_FOUND") {
-            console.log("TODO HERE MISSING", { srcPath, depName });
+            const srcPkg = getLastPackage(srcPath);
+            const isAllowed = srcPkg && (allowMissing[srcPkg] || [])
+              .some((pkg) => depName === pkg || depName.startsWith(`${pkg}/`));
+
+            if (isAllowed) {
+              return null;
+            }
           }
 
           // Convert to more useful error.
           throw new Error(`Encountered resolution error in ${srcPath} for ${depName}: ${err}`);
         });
     }))
-    // Remove core standard Node.js libraries (`path`, etc.)
-    .then((allPaths) => allPaths.filter(({ depPath }, i) => depPath !== depNames[i]));
+    // Post-resolution processing.
+    .then((allPaths) => allPaths
+      // Remove empty names (e.g., from allowed missing modules).
+      .filter(Boolean)
+      // Remove core standard Node.js libraries (`path`, etc.)
+      .filter(({ depPath }, i) => depPath !== depNames[i])
+    );
 
   // Extract to different types of paths.
   const depPaths = depObjs.map(({ depPath }) => depPath);
@@ -124,7 +136,7 @@ const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } =
     // Add all needed package.json files
     allPkgPaths,
     // Add dependencies from recursion.
-    await _recurseDeps({ depPaths, ignores, tracedDepPaths })
+    await _recurseDeps({ depPaths, ignores, allowMissing, tracedDepPaths })
   );
 
   // Make unique and return.
@@ -141,9 +153,9 @@ const traceFile = async ({ srcPath, ignores = [], tracedDepPaths = new Set() } =
  * @param {Set}           opts.tracedDepPaths tracked dependencies
  * @returns {Promise<Array<string>>}          list of absolute paths to on-disk dependencies
  */
-const traceFiles = async ({ srcPaths, ignores = [], tracedDepPaths = new Set() } = {}) => {
+const traceFiles = async ({ srcPaths, ignores = [], allowMissing = {}, tracedDepPaths = new Set() } = {}) => {
   // Recurse all source files.
-  const resolvedDepPaths = await _recurseDeps({ srcPaths, ignores, tracedDepPaths });
+  const resolvedDepPaths = await _recurseDeps({ srcPaths, ignores, allowMissing, tracedDepPaths });
 
   // Make unique and return.
   return uniq(resolvedDepPaths);

--- a/lib/trace.js
+++ b/lib/trace.js
@@ -33,7 +33,9 @@ const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, allowMissing, tr
 
       // Recurse.
       // eslint-disable-next-line no-use-before-define
-      const recursedDepPaths = await traceFile({ srcPath: depPath, ignores, allowMissing, tracedDepPaths });
+      const recursedDepPaths = await traceFile(
+        { srcPath: depPath, ignores, allowMissing, tracedDepPaths }
+      );
 
       // Aggregate.
       resolvedDepPaths = resolvedDepPaths.concat(recursedDepPaths);
@@ -60,7 +62,12 @@ const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, allowMissing, tr
  * @param {Set}           opts.tracedDepPaths tracked dependencies
  * @returns {Promise<Array<string>>}          list of absolute paths to on-disk dependencies
  */
-const traceFile = async ({ srcPath, ignores = [], allowMissing = {}, tracedDepPaths = new Set() } = {}) => {
+const traceFile = async ({
+  srcPath,
+  ignores = [],
+  allowMissing = {},
+  tracedDepPaths = new Set()
+} = {}) => {
   if (!srcPath) {
     throw new Error("Empty source file path");
   }
@@ -153,7 +160,12 @@ const traceFile = async ({ srcPath, ignores = [], allowMissing = {}, tracedDepPa
  * @param {Set}           opts.tracedDepPaths tracked dependencies
  * @returns {Promise<Array<string>>}          list of absolute paths to on-disk dependencies
  */
-const traceFiles = async ({ srcPaths, ignores = [], allowMissing = {}, tracedDepPaths = new Set() } = {}) => {
+const traceFiles = async ({
+  srcPaths,
+  ignores = [],
+  allowMissing = {},
+  tracedDepPaths = new Set()
+} = {}) => {
   // Recurse all source files.
   const resolvedDepPaths = await _recurseDeps({ srcPaths, ignores, allowMissing, tracedDepPaths });
 

--- a/test/lib/package.spec.js
+++ b/test/lib/package.spec.js
@@ -1,0 +1,36 @@
+"use strict";
+
+const { normalize } = require("path");
+
+const { getPackages } = require("../../lib/package");
+
+describe("lib/package", () => {
+  describe("getPackages", () => {
+    it("handles empty input", () => {
+      expect(getPackages()).to.eql([]);
+      expect(getPackages([])).to.eql([]);
+      expect(getPackages(null)).to.eql([]);
+    });
+
+    it("handles no modules", () => {
+      expect(getPackages(normalize(""))).to.eql([]);
+      expect(getPackages(normalize("src"))).to.eql([]);
+      expect(getPackages(normalize("path/to/node_modules"))).to.eql([]);
+      expect(getPackages(normalize("path/to/node_modules/@scope"))).to.eql([]);
+    });
+
+    it("handles single modules", () => {
+      expect(getPackages(normalize("node_modules/bar"))).to.eql(["bar"]);
+      expect(getPackages(normalize("path/to/node_modules/@scope/foo"))).to.eql(["@scope/foo"]);
+    });
+
+    it("handles nested modules", () => {
+      expect(getPackages(normalize("node_modules/bar/node_modules/@scope/nested-bar")))
+        .to.eql(["bar", "@scope/nested-bar"]);
+      expect(getPackages(normalize(
+        "path/to/node_modules/@scope/foo/node_modules/three/node_modules/four/node_modules"
+      )))
+        .to.eql(["@scope/foo", "three", "four"]);
+    });
+  });
+});

--- a/test/lib/package.spec.js
+++ b/test/lib/package.spec.js
@@ -2,13 +2,12 @@
 
 const { normalize } = require("path");
 
-const { getPackages } = require("../../lib/package");
+const { getPackages, getLastPackage } = require("../../lib/package");
 
 describe("lib/package", () => {
   describe("getPackages", () => {
     it("handles empty input", () => {
       expect(getPackages()).to.eql([]);
-      expect(getPackages([])).to.eql([]);
       expect(getPackages(null)).to.eql([]);
     });
 
@@ -31,6 +30,35 @@ describe("lib/package", () => {
         "path/to/node_modules/@scope/foo/node_modules/three/node_modules/four/node_modules"
       )))
         .to.eql(["@scope/foo", "three", "four"]);
+    });
+  });
+
+  describe("getLastPackage", () => {
+    it("handles empty input", () => {
+      expect(getLastPackage()).to.eql(null);
+      expect(getLastPackage(null)).to.eql(null);
+    });
+
+    it("handles no modules", () => {
+      expect(getLastPackage(normalize(""))).to.eql(null);
+      expect(getLastPackage(normalize("src"))).to.eql(null);
+      expect(getLastPackage(normalize("path/to/node_modules"))).to.eql(null);
+      expect(getLastPackage(normalize("path/to/node_modules/@scope"))).to.eql(null);
+    });
+
+
+    it("handles single modules", () => {
+      expect(getLastPackage(normalize("node_modules/bar"))).to.eql("bar");
+      expect(getLastPackage(normalize("path/to/node_modules/@scope/foo"))).to.eql("@scope/foo");
+    });
+
+    it("handles nested modules", () => {
+      expect(getLastPackage(normalize("node_modules/bar/node_modules/@scope/nested-bar")))
+        .to.eql("@scope/nested-bar");
+      expect(getLastPackage(normalize(
+        "path/to/node_modules/@scope/foo/node_modules/three/node_modules/four/node_modules"
+      )))
+        .to.eql("four");
     });
   });
 });

--- a/test/lib/trace.spec.js
+++ b/test/lib/trace.spec.js
@@ -525,13 +525,12 @@ describe("lib/trace", () => {
       ]));
     });
 
-    // TODO(19): Reenable and implement support.
-    it.skip("handles try/catch missing requires", async () => {
+    it("handles try/catch missing requires", async () => {
       mock({
         "hi.js": `
           require("one");
 
-          const { aFunction } = require("nested-trycatch-require");
+          const { aFunction } = require("nested-first-level");
           const { aFile } = require("nested-trycatch-requireresolve");
         `,
         node_modules: {
@@ -540,8 +539,6 @@ describe("lib/trace", () => {
               main: "index.js"
             }),
             "index.js": `
-              require("doesnt-exist");
-
               module.exports = {
                 one: () => "one",
                 two: () => require("two").two
@@ -558,6 +555,20 @@ describe("lib/trace", () => {
               };
             `
           },
+          "nested-first-level": {
+            "package.json": stringify({
+              main: "lib/index.js"
+            }),
+            lib: {
+              "index.js": `
+                const { aFunction } = require("nested-trycatch-require");
+
+                module.exports = {
+                  aFunction
+                };
+              `
+            }
+          },
           "nested-trycatch-require": {
             "package.json": stringify({
               main: "index.js"
@@ -565,7 +576,7 @@ describe("lib/trace", () => {
             "index.js": `
               let aFunction;
               try {
-                aFunction = () => import("doesnt-exist");
+                aFunction = () => import("doesnt-exist/with/path.js");
               } catch (err) {
                 aFunction = () => null;
               }
@@ -592,8 +603,20 @@ describe("lib/trace", () => {
       });
 
       expect(await traceFile({
-        srcPath: "hi.js"
+        srcPath: "hi.js",
+        allowMissing: {
+          "nested-trycatch-require": [
+            "doesnt-exist"
+          ],
+          "nested-trycatch-requireresolve": [
+            "also-doesnt-exist"
+          ]
+        }
       })).to.eql(fullPath([
+        "node_modules/nested-first-level/lib/index.js",
+        "node_modules/nested-first-level/package.json",
+        "node_modules/nested-trycatch-require/index.js",
+        "node_modules/nested-trycatch-require/package.json",
         "node_modules/nested-trycatch-requireresolve/index.js",
         "node_modules/nested-trycatch-requireresolve/package.json",
         "node_modules/one/index.js",

--- a/test/lib/trace.spec.js
+++ b/test/lib/trace.spec.js
@@ -524,6 +524,70 @@ describe("lib/trace", () => {
         "node_modules/two/package.json"
       ]));
     });
+
+    it("handles try/catch missing requires", async () => {
+      mock({
+        "hi.js": `
+          require("one");
+
+          let noFunction;
+          try {
+            noFunction = () => import("doesnt-exist");
+          } catch (err) {}
+
+          const aFile = require("nested-trycatch-requireresolve");
+        `,
+        node_modules: {
+          one: {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              require("doesnt-exist");
+
+              module.exports = {
+                one: () => "one",
+                two: () => require("two").two
+              };
+            `
+          },
+          two: {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              module.exports = {
+                two: () => "two"
+              };
+            `
+          },
+          "nested-trycatch-requireresolve": {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              let noFile = null;
+              try {
+                noFile = require.resolve("also-doesnt-exist");
+              } catch (err) {}
+
+              module.exports = { noFile };
+            `
+          }
+        }
+      });
+
+      expect(await traceFile({
+        srcPath: "hi.js"
+      })).to.eql(fullPath([
+        "node_modules/nested-trycatch-requireresolve/index.js",
+        "node_modules/nested-trycatch-requireresolve/package.json",
+        "node_modules/one/index.js",
+        "node_modules/one/package.json",
+        "node_modules/two/index.js",
+        "node_modules/two/package.json"
+      ]));
+    });
   });
 
   describe("traceFiles", () => {

--- a/test/lib/trace.spec.js
+++ b/test/lib/trace.spec.js
@@ -597,8 +597,6 @@ describe("lib/trace", () => {
               module.exports = { noFile };
             `
           }
-
-
         }
       });
 
@@ -629,33 +627,9 @@ describe("lib/trace", () => {
     it("still errors on missing imports in a catch", async () => {
       mock({
         "hi.js": `
-          require("one");
-
           const { aFunction } = require("nested-first-level");
-          const { aFile } = require("nested-trycatch-requireresolve");
         `,
         node_modules: {
-          one: {
-            "package.json": stringify({
-              main: "index.js"
-            }),
-            "index.js": `
-              module.exports = {
-                one: () => "one",
-                two: () => require("two").two
-              };
-            `
-          },
-          two: {
-            "package.json": stringify({
-              main: "index.js"
-            }),
-            "index.js": `
-              module.exports = {
-                two: () => "two"
-              };
-            `
-          },
           "nested-first-level": {
             "package.json": stringify({
               main: "lib/index.js"
@@ -684,22 +658,7 @@ describe("lib/trace", () => {
 
               module.exports = { aFunction };
             `
-          },
-          "nested-trycatch-requireresolve": {
-            "package.json": stringify({
-              main: "index.js"
-            }),
-            "index.js": `
-              let noFile = null;
-              try {
-                noFile = require.resolve("also-doesnt-exist");
-              } catch (err) {}
-
-              module.exports = { noFile };
-            `
           }
-
-
         }
       });
 
@@ -709,9 +668,6 @@ describe("lib/trace", () => {
           "nested-first-level": [
             // This won't be a permitted missing because only `nested-trycatch-require` is checked.
             "doesnt-exist"
-          ],
-          "nested-trycatch-requireresolve": [
-            "also-doesnt-exist"
           ]
         }
       })).to.be.rejectedWith(

--- a/test/lib/trace.spec.js
+++ b/test/lib/trace.spec.js
@@ -626,7 +626,98 @@ describe("lib/trace", () => {
       ]));
     });
 
-    it("still errors on missing imports in a catch"); // TODO(19)
+    it("still errors on missing imports in a catch", async () => {
+      mock({
+        "hi.js": `
+          require("one");
+
+          const { aFunction } = require("nested-first-level");
+          const { aFile } = require("nested-trycatch-requireresolve");
+        `,
+        node_modules: {
+          one: {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              module.exports = {
+                one: () => "one",
+                two: () => require("two").two
+              };
+            `
+          },
+          two: {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              module.exports = {
+                two: () => "two"
+              };
+            `
+          },
+          "nested-first-level": {
+            "package.json": stringify({
+              main: "lib/index.js"
+            }),
+            lib: {
+              "index.js": `
+                const { aFunction } = require("nested-trycatch-require");
+
+                module.exports = {
+                  aFunction
+                };
+              `
+            }
+          },
+          "nested-trycatch-require": {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              let aFunction;
+              try {
+                aFunction = () => import("doesnt-exist/with/path.js");
+              } catch (err) {
+                aFunction = () => null;
+              }
+
+              module.exports = { aFunction };
+            `
+          },
+          "nested-trycatch-requireresolve": {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              let noFile = null;
+              try {
+                noFile = require.resolve("also-doesnt-exist");
+              } catch (err) {}
+
+              module.exports = { noFile };
+            `
+          }
+
+
+        }
+      });
+
+      await expect(traceFile({
+        srcPath: "hi.js",
+        allowMissing: {
+          "nested-first-level": [
+            // This won't be a permitted missing because only `nested-trycatch-require` is checked.
+            "doesnt-exist"
+          ],
+          "nested-trycatch-requireresolve": [
+            "also-doesnt-exist"
+          ]
+        }
+      })).to.be.rejectedWith(
+        /Encountered resolution error in .*nested-trycatch-require.* for doesnt-exist.*/
+      );
+    });
   });
 
   describe("traceFiles", () => {

--- a/test/lib/trace.spec.js
+++ b/test/lib/trace.spec.js
@@ -525,17 +525,14 @@ describe("lib/trace", () => {
       ]));
     });
 
-    it("handles try/catch missing requires", async () => {
+    // TODO(19): Reenable and implement support.
+    it.skip("handles try/catch missing requires", async () => {
       mock({
         "hi.js": `
           require("one");
 
-          let noFunction;
-          try {
-            noFunction = () => import("doesnt-exist");
-          } catch (err) {}
-
-          const aFile = require("nested-trycatch-requireresolve");
+          const { aFunction } = require("nested-trycatch-require");
+          const { aFile } = require("nested-trycatch-requireresolve");
         `,
         node_modules: {
           one: {
@@ -561,6 +558,21 @@ describe("lib/trace", () => {
               };
             `
           },
+          "nested-trycatch-require": {
+            "package.json": stringify({
+              main: "index.js"
+            }),
+            "index.js": `
+              let aFunction;
+              try {
+                aFunction = () => import("doesnt-exist");
+              } catch (err) {
+                aFunction = () => null;
+              }
+
+              module.exports = { aFunction };
+            `
+          },
           "nested-trycatch-requireresolve": {
             "package.json": stringify({
               main: "index.js"
@@ -574,6 +586,8 @@ describe("lib/trace", () => {
               module.exports = { noFile };
             `
           }
+
+
         }
       });
 
@@ -588,6 +602,8 @@ describe("lib/trace", () => {
         "node_modules/two/package.json"
       ]));
     });
+
+    it("still errors on missing imports in a catch"); // TODO(19)
   });
 
   describe("traceFiles", () => {

